tests/integration_lma/test_lma_bundle.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import os
+import shutil
+import tempfile
+
+import pytest
+from git import Repo
+
+log = logging.getLogger(__name__)
+
```

Comment:
This is a comment on the whole file. If I understand correctly, this appears to be a pytest test which clones the repo we want and runs pytest against it. Why are we doing this in a test. Is it not possible to add the following to tox.ini?
```
git clone ...
pytest ...
```
Besides most likely leading to bugs, this is just incredibly confusing for no apparent benefit. If there is a reason here that I don't know about, please at least put some comments at the top of the file explaining the need for it.

Reply:
I had tox taking most of the burden before switching to this pattern.
Apart from git clone and pytest, there is also `charm_under_test = await ops_test.build_charm(".")`, and to replace that I would have to call charmcraft from tox. [It was ugly](https://github.com/canonical/alertmanager-operator/pull/34/commits/feb05f9d614fcb9defd2ff46cbf0a9f483a49cb0).

Reply:
Two points here:
1. I don't think that was ugly.
2. Why can't you still use ops_test.build_charm in the integration tests? You are calling pytest still.

Reply:
**1**
I think you overlooked [this line](https://github.com/canonical/alertmanager-operator/blob/feb05f9d614fcb9defd2ff46cbf0a9f483a49cb0/tox.ini#L103):
`tmpdir = /tmp/pseudo_random/scf3u8/in_lieu_of_mktemp_d`

Also for that to work I needed to roll my own `render_bundle.py`.
Also, for cleanup, there is an `rm -rf` in there, which is ugly on its own.

**2**
Indeed, now I build the charm inside the python code. But that also means that I need to call pytest from within the python code, because tox wouldn't know the path of the *.charm (unless I add even more ugliness).
Basically it's either split the test logic into tox + py + ugliness, or keep it all inside the py. I think the keeping it all inside the py makes the most sense.

Reply:
Sorry to be a thorn in your side here but if this is going to end up in every repo I feel I must.

1. I don't think using a tmpdir is an issue. Are you concerned that you don't get the same safety as the python tempfile library? If so, the command `mktemp` is almost certainly in the github containers and does the same thing.

Haven't you made a render_bundle.py already?

`rm` is how you delete files on linux. I'm not sure what the issue is with running it

2. I'm not really following here. What do you mean tox wouldn't know the path? If pytest builds the charm and runs tests on it, why would tox need to know anything about it?

Reply:
In my experience, tmpdirs are far better at avoiding dirty-state bugs, and markedly worse for debuggability. Which trade off  to take here depends on the likelihood and possible severity of dirty-state bugs.

Reply:
1
Good point about storing the variable. I probably would make a shell script or python script.

>> Besides most likely leading to bugs, this is just incredibly confusing for no apparent benefit.
> This argument is vague so please defend it, otherwise I will resolve :)

It is very confusing the way it currently works. We are calling a pytest test sweet that has a single instance of a test. That test deploys a bundle then initiates another call of pytest which itself has a bunch of tests. What is the significance of the results of the outer test vs the inner tests? Why do we have `assert 5 == 7`? Does this mean the outer test always fails?

When looking at this design there are just so many questions and from the outside looking in it looks like we are using pytest to run code rather than run tests.

2
I think we are in agreement on this thread. Tox does not need to build the charms/bundle.

Reply:
> Why do we have assert 5 == 7?

Oops! It was for sanity check.

> it looks like we are using pytest to run code rather than run tests.

It's a bit more subtle: the "outer pytest" also builds the charm locally which is worthy of being under pytest, and, yes, does some file ops.
This goes back to the tradeoff between:
- Everything is inside pytest py code, and enjoys the built-in ops_test context, and works exactly the same locally and on ghwf; vs
- Use a shell script (or [equivalent](https://github.com/canonical/alertmanager-operator/blob/feb05f9d614fcb9defd2ff46cbf0a9f483a49cb0/tox.ini) of) to git clone, sg lxd charmcraft pack etc., splitting test logic across tox -> shell script.